modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java

```diff
@@ -166,45 +171,110 @@ private void ensureOpen() {
     }
 
     private void reopenStreamOrFail(IOException e) throws IOException {
-        if (purpose == OperationPurpose.REPOSITORY_ANALYSIS) {
-            logger.warn(() -> format("""
-                failed reading [%s/%s] at offset [%s]""", blobStore.bucket(), blobKey, start + currentOffset), e);
-            throw e;
+        maybeRecordAndLogForRetry("reading", e);
+        maybeAbort(currentStream);
+        IOUtils.closeWhileHandlingException(currentStream);
+
+        maybeDelaySafely();
+        openStreamWithRetry();
+    }
+
+    private <T extends Exception> void maybeRecordAndLogForRetry(String action, T e) throws T {
+        if (shouldRetry() == false) {
+            final var finalException = addSuppressedExceptions(e);
+            logForRetry(Level.WARN, action, finalException);
+            throw finalException;
         }
 
-        final int maxAttempts = blobStore.getMaxRetries() + 1;
+        attempt += 1;
```

Comment:
Would it be possible emit apm metric always when we retry? 
I'd like to have a sense how many retries do we do (avg, count)

Reply:
It could a useful metric specific to s3. I can work on it in a separate follow-up. Thanks for the idea.